[Examples] InstructPix2Pix instruct training script

sayakpaul opened the pull request:
Closes https://github.com/huggingface/diffusers/issues/2288.

I opted for using WandB tables instead of just images, as I think tables here are better suited. 

[Here's](https://wandb.ai/sayakpaul/uncategorized/reports/Weave-test_table-23-02-24-11-03-40---VmlldzozNjMyODk1) how they would look like. 

## Command to launch an experiment (with `wandb` logging)

```bash
accelerate launch --mixed_precision="fp16" train_instruct_pix2pix.py \
 --pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5 \
 --dataset_name=sayakpaul/instructpix2pix-1000-samples \
 --use_ema \
 --enable_xformers_memory_efficient_attention \
 --resolution=256 --random_flip \
 --train_batch_size=4 --gradient_accumulation_steps=4 --gradient_checkpointing \
 --max_train_steps=15000 \
 --checkpointing_steps=5000 --checkpoints_total_limit=1 \
 --learning_rate=5e-05 --max_grad_norm=1 --lr_warmup_steps=0 \
 --conditioning_dropout_prob=0.05 \
 --mixed_precision=fp16 \
 --val_image_url="https://hf.co/datasets/diffusers/diffusers-images-docs/resolve/main/mountain.png" \
 --validation_prompt="make the mountains snowy" \
 --seed=42 \
 --report_to=wandb 
```

Run ongoing here: https://wandb.ai/sayakpaul/instruct-pix2pix/runs/kjmrkjop

## Testing machine 

### `diffusers-cli env`

- `diffusers` version: 0.14.0.dev0
- Platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.10
- Python version: 3.8.16
- PyTorch version (GPU?): 1.13.1+cu116 (True)
- Huggingface_hub version: 0.12.0
- Transformers version: 4.26.1
- Accelerate version: 0.16.0
- xFormers version: 0.0.16
- Using GPU in script?: Yes (A100 40 GB)
- Using distributed or parallel set-up in script?: No

## TODO

- [x] Get the script to work
- [ ] ~Run experiments to find reasonable hyperparameters~ (for now)
- [x] Add requirements.txt 
- [x] Fill in the README
- [ ] Add an entry to the training section in the docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2478). All of your documentation changes will be reflected on that endpoint.

sayakpaul commented:
@patrickvonplaten I ran two experiments with the current state of the training script and I will post my analysis soon.

sayakpaul commented:
@patrickvonplaten @patil-suraj so I ran another experiment with:

```bash
accelerate launch --mixed_precision="fp16" train_instruct_pix2pix.py \
--pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5 \
--dataset_name=sayakpaul/instructpix2pix-1000-samples \
--use_ema \
--enable_xformers_memory_efficient_attention \
--resolution=512 --random_flip \
--train_batch_size=2 --gradient_accumulation_steps=4 --gradient_checkpointing \
--max_train_steps=15000 \
--checkpointing_steps=5000 --checkpoints_total_limit=1 \
--learning_rate=5e-05 --lr_warmup_steps=0 \
--conditioning_dropout_prob=0.05 \
--mixed_precision=fp16 \
--val_image_url="https://hf.co/datasets/diffusers/diffusers-images-docs/resolve/main/mountain.png" \
--validation_prompt="make the mountains snowy" \
--seed=42 \
--report_to=wandb 
```

The final trained model is here: https://huggingface.co/sayakpaul/instruct-pix2pix.

I ran inference with this model:

```python
import PIL
import requests
import torch
from diffusers import StableDiffusionInstructPix2PixPipeline

model_id = "sayakpaul/instruct-pix2pix"
pipe = StableDiffusionInstructPix2PixPipeline.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
generator = torch.Generator("cuda").manual_seed(0)

url = "https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/test_pix2pix_4.png"


def download_image(url):
    image = PIL.Image.open(requests.get(url, stream=True).raw)
    image = PIL.ImageOps.exif_transpose(image)
    image = image.convert("RGB")
    return image


image = download_image(url)

prompt = "wipe out the lake"

for num_inference_steps in [20, 50, 100]:
    for image_guidance_scale in [1.0, 1.5, 2.0]:
        for guidance_scale in [5, 6, 7, 8, 10]:
            edited_image = pipe(prompt, 
                image=image, 
                num_inference_steps=num_inference_steps, 
                image_guidance_scale=image_guidance_scale, 
                guidance_scale=guidance_scale,
                generator=generator,
            ).images[0]
            edited_image.save(
                f"lake_inf_steps@{num_inference_steps}_imgs@{image_guidance_scale}_gs@{guidance_scale}.png"
            )
```

Here are the results:
[inference_images.zip](https://github.com/huggingface/diffusers/files/10857126/inference_images.zip)

Note that the sample I used for inference is actually present in the [mini training dataset](https://huggingface.co/datasets/sayakpaul/instructpix2pix-1000-samples) I used for the example script. 

You would notice that the results are not up to the mark. I suspect:

* The network is still under-trained.
* Lack of training instances. 

Both these aspects can be improved with more rigorous hyperparameter tuning. Regardless of those, I believe it would be great to review the script's implementation details if you have time.

Arminias commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,991 @@
+    def preprocess_train(examples):
+        original_images = [image.convert("RGB") for image in examples[original_image_column]]
+        edited_images = [image.convert("RGB") for image in examples[edited_image_column]]
+        examples["original_pixel_values"] = [train_transforms(image) for image in original_images]
```
Noticed weird results (output images from pipeline flipping randomly) during testing when the `random_flip` option was enabled. Cause is ultimately here I think: The way the `train_transforms` are called here does unfortunately not ensure that corresponding input and output images are flipped at the same time.

sayakpaul replied:
Thank you for your observing! You're indeed correct, I will push a fix.

sayakpaul replied:
This should now be fixed. Thanks again for pointing it out.

patrickvonplaten commented:
Sorry for the delay here - going through it now

patrickvonplaten commented on examples/instruct_pix2pix/requirements.txt:
```diff
@@ -0,0 +1,7 @@
+datasets
+ftfy
+tensorboard
+Jinja2
```
```suggestion
```

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+    vae = AutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae", revision=args.revision)
+    unet = UNet2DConditionModel.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="unet", revision=args.non_ema_revision
+    )
```
```suggestion
    )
    unet.register_to_config(in_channel=8)
    conv_in = nn.Linear(..., bias=False)
    conv_in.weight[:, :4, :, :].copy_(unet.conv_in.weight)
    unet.conv_in = conv_in
```

Think just those three lines would be nicer than to create a whole new function

sayakpaul replied:
Thanks! This looks more elegant and indeed cleaner. One cannot do `conv_in.weight[:, :4, :, :].copy_(unet.conv_in.weight)` because in-place operations cannot be performed on leaf variables that require grads. So, if one puts the in-place ops within a `torch.no_grad()` context it should work. 

Full working snippet:

```python
out_channels = unet.conv_in.out_channels
in_channels = 8

with torch.no_grad():
    new_conv_in = nn.Conv2d(
        in_channels, out_channels, unet.conv_in.kernel_size, unet.conv_in.stride, unet.conv_in.padding
    )
    new_conv_in.weight.zero_()
    new_conv_in.weight[:, :4, :, :].copy_(unet.conv_in.weight)
    unet.conv_in = new_conv_in
```

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+
+    if accelerator.is_main_process:
+        logger.info("Initializing the InstructPix2Pix UNet from the pretrained UNet.")
+        instruct_pix2pix_unet = initialize_unet(unet, instruct_pix2pix_unet)
```
Think would prefer to just not add this function here think it complicates thinks a bit too much

sayakpaul replied:
Should be fixed now.

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+
+    # Freeze vae and text_encoder
+    vae.requires_grad_(False)
+    text_encoder.requires_grad_(False)
```
text encoder is never trained for instruct pix2pix?

sayakpaul replied:
Nope. But feel free to verify this from Section 3.2 of the [original paper](https://arxiv.org/pdf/2211.09800.pdf).

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+        ema_unet = EMAModel(ema_unet.parameters(), model_cls=UNet2DConditionModel, model_config=ema_unet.config)
+
+    # Remove the `unet` as we don't need it.
+    del unet
```
I'd only ever instantiate one unet to begin with and use the name `unet` throughout the script

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+                logger.warn(
+                    "xFormers 0.0.16 cannot be used for training in some GPUs. If you observe problems during training, please update xFormers to at least 0.0.17. See https://huggingface.co/docs/diffusers/main/en/optimization/xformers for more details."
+                )
+            instruct_pix2pix_unet.enable_xformers_memory_efficient_attention()
```
```suggestion
            unet.enable_xformers_memory_efficient_attention()
```

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+                # Conditioning dropout to support classifier-free guidance during inference. For more details
+                # check out the section 3.2.1 of the original paper https://arxiv.org/abs/2211.09800.
+                if args.conditioning_dropout_prob is not None:
+                    random_p = torch.rand(bsz, device=latents.device)
```
```suggestion
                    random_p = torch.rand(bsz, device=latents.device, generator=generator)
```

I'd maybe a generator here in the beginning of the script to make it more reproducible

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+                    prompt_mask = random_p < 2 * args.conditioning_dropout_prob
+                    prompt_mask = prompt_mask.reshape(bsz, 1, 1)
+                    # Final text conditioning.
+                    null_conditioning = text_encoder(tokenize_captions([NULL_PROMPT]).to(accelerator.device))[0]
```
```suggestion
                    null_conditioning = text_encoder(tokenize_captions([""]).to(accelerator.device))[0]
```

patrickvonplaten reviewed (commented):
Looks very cool! Left some nits, the main thing for me is not not instantiate multiple unets in the beginning just to extent one weight. Think we can solve this easier in just 4 lines witout a function. Also think such as the layer name or null weight have to be defined as constants in the beginning - think we can just do this "in-line" which helps readability. 

Totally ok for me to merge without super promising results yet, but can we maybe have one nice training run and W&B link?

patrickvonplaten commented on examples/instruct_pix2pix/README.md:
```diff
@@ -0,0 +1 @@
+# Training InstructPix2Pix
```
Let's maybe add a nice description here and add a link such as this one: https://wandb.ai/sayakpaul/uncategorized/reports/Weave-test_table-23-02-24-11-03-40---VmlldzozNjMyODk1 

We can then let the community try to improve the script on larger training runs

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1018 @@
+
+    # Create EMA for the unet.
+    if args.use_ema:
+        ema_unet = UNet2DConditionModel.from_config(instruct_pix2pix_config)
```
Deepcopy? 

I followed what we do for the other examples.

sayakpaul replied:
Ah got it. See if the latest updates make sense.

sayakpaul commented:
@patrickvonplaten thanks for helping to clean the script and make it better readability-wise. I think I have addressed all your comments. 

**Updates**:

* Training going on here: https://wandb.ai/sayakpaul/instruct-pix2pix/runs/lgstrwko on our small dataset: https://huggingface.co/datasets/fusing/instructpix2pix-1000-samples
* Updated the README to communicate where we currently stand clearly. I will update it with a W&B run link once the current training run is done. 

> We can then let the community try to improve the script on larger training runs

Actually, this script can be extended to longer training runs and larger datasets without any code-changes. We have tried incorporating everything that makes InstructPix2Pix work the way it works. I will make sure that the README reflects it. 

@williamberman feel free to review when you get a moment.

patrickvonplaten commented:
Perf! Script looks good for merge for me!

patrickvonplaten reviewed (approved):
Would be nice if you could give it a quick review @williamberman

sayakpaul commented:
@williamberman I am facing something weird here. 

The final pipeline inference fails **_mid-way_**:

```bash
  File "train_instruct_pix2pix.py", line 1010, in <module>
    main()
  File "train_instruct_pix2pix.py", line 987, in main
    pipeline(
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py", line 380, in __call__
    image = self.decode_latents(latents)
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py", line 643, in decode_latents
    image = self.vae.decode(latents).sample
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/diffusers/models/autoencoder_kl.py", line 185, in decode
    decoded = self._decode(z).sample
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/diffusers/models/autoencoder_kl.py", line 171, in _decode
    z = self.post_quant_conv(z)
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 463, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/opt/conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 459, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Input type (float) and bias type (c10::Half) should be the same
```

Here's a command to reproduce it quickly:

```bash
accelerate launch --mixed_precision="fp16" train_instruct_pix2pix.py 
  --pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5 
  --dataset_name=sayakpaul/instructpix2pix-1000-samples 
  --enable_xformers_memory_efficient_attention 
  --resolution=512 --random_flip --train_batch_size=2 
  --gradient_accumulation_steps=4 --gradient_checkpointing 
  --max_train_steps=2 --max_train_samples 10 
  --checkpointing_steps=1 
  --checkpoints_total_limit=1 
  --learning_rate=5e-05 --lr_warmup_steps=0 
  --conditioning_dropout_prob=0.05 --mixed_precision=fp16 
  --val_image_url="https://hf.co/datasets/diffusers/diffusers-images-docs/resolve/main/mountain.png" 
  --validation_prompt="make the mountains snowy" 
  --seed=42 --report_to=wandb
```

Outputs of the print statements I have added for debugging:

```bash
Pipeline device: cuda:0
UNet: cuda:0 Text Encoder: cuda:0 VAE: cuda:0
UNet: torch.float16 Text Encoder: torch.float16 VAE: torch.float16
```

When you get a chance, could you look into it once?


Cc: @patrickvonplaten

patrickvonplaten commented on examples/instruct_pix2pix/train_instruct_pix2pix.py:
```diff
@@ -0,0 +1,1010 @@
+                    f"UNet: {pipeline.unet.dtype} Text Encoder: {pipeline.text_encoder.dtype} VAE: {pipeline.vae.dtype}"
+                )
+
+                edited_images.append(
```
```suggestion
                edited_images.append(
                    with torch.autocast("cuda"):
```

Should solve the problem. See reasoning here: https://github.com/huggingface/diffusers/issues/2568

sayakpaul replied:
Thank you!
 
But why it would be needed when the dtypes and devices are uniform across the sub-models?

sayakpaul replied:
@patrickvonplaten, it still fails with the change.

sayakpaul commented:
@patrickvonplaten issue resolved. Here's what I had to do following https://github.com/huggingface/diffusers/pull/2631:

* I unwrapped the `vae` and `text_encoder` along with `unet`. 
* I wrapped the final inference block within autocast as you suggested. We should be good now :)

patrickvonplaten commented:
> @patrickvonplaten issue resolved. Here's what I had to do following #2631:
> 
> * I unwrapped the `vae` and `text_encoder` along with `unet`.
> * I wrapped the final inference block within autocast as you suggested. We should be good now :)

Ah interesting! Thanks for the pointer. Pr is good to merge for me :-) Nice job
